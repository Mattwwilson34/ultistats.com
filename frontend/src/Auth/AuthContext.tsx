import React, { createContext, useEffect, useState } from 'react'
import { auth } from '../../../firebase/firebase'
import { type User } from 'firebase/auth'
import { CircularProgress } from '@mui/material'

interface AuthContextType {
  currentUser: User | null
}

const AuthContext = createContext<AuthContextType>({ currentUser: null })

const AuthProvider = ({
  children,
}: {
  children: React.ReactNode
}): JSX.Element => {
  const [currentUser, setCurrentUser] = useState<User | null>(null)
  const [pending, setPending] = useState<boolean>(true)

  useEffect(() => {
    auth.onAuthStateChanged((user) => {
      setCurrentUser(user)
      setPending(false)
    })
  }, [currentUser])

  if (pending) {
    return (
      <>
        <CircularProgress />
      </>
    )
  }
  return (
    <AuthContext.Provider value={{ currentUser }}>
      {children}
    </AuthContext.Provider>
  )
}

export { AuthProvider, AuthContext }
