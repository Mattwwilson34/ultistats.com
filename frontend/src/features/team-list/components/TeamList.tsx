import * as React from 'react'
import Box from '@mui/material/Box'
import List from '@mui/material/List'
import ListItem from '@mui/material/ListItem'
import ListItemButton from '@mui/material/ListItemButton'
import AddIcon from '@mui/icons-material/Add'
import ListItemText from '@mui/material/ListItemText'
import ListItemIcon from '@mui/material/ListItemIcon'
import Divider from '@mui/material/Divider'
import TeamListItem from './TeamListItem'
import { uuidv4 } from '@firebase/util'

interface TeamListProps {
  teams: string[] | null
}

export default function TeamList({ teams }: TeamListProps): React.ReactElement {
  return (
    <Box sx={{ width: '100%', bgcolor: 'background.paper' }}>
      <nav aria-label="secondary mailbox folders">
        <List>
          <ListItem disablePadding>
            <ListItemButton>
              <ListItemText primary="New Team" />
              <ListItemIcon>
                <AddIcon />
              </ListItemIcon>
            </ListItemButton>
          </ListItem>
        </List>
      </nav>
      <Divider />
      <nav aria-label="secondary mailbox folders">
        {teams == null
          ? null
          : teams.map((team) => {
              return (
                <List key={uuidv4()}>
                  <TeamListItem teamName={team} />
                </List>
              )
            })}
      </nav>
    </Box>
  )
}
