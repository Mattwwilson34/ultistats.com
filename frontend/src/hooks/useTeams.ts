import { useQuery, type UseQueryResult } from 'react-query'
import { db } from '../../../firebase/firebase'
import { collection, query, where, getDocs } from 'firebase/firestore'
import { type Team } from '../types/teams'

type Teams = Team[]

const getTeams = async (userUid: string): Promise<Teams> => {
  const teams: Teams = []
  const teamsRef = collection(db, 'teams')
  const q = query(teamsRef, where('userUid', '==', userUid))
  const querySnapshot = await getDocs(q)
  querySnapshot.forEach((doc) => {
    // Fix: teamName is undefined in this scope. Moving on with dev for now.
    const team = { ...doc.data() } as Team // eslint-disable-line @typescript-eslint/consistent-type-assertions

    teams.push(team)
  })
  return teams
}

const useTeams = (userId: string): UseQueryResult<Teams> => {
  return useQuery(['teams', userId], async () => await getTeams(userId))
}
//
export default useTeams
