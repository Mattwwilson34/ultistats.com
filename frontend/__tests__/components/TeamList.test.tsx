import React from 'react'
import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import TeamList from '../../src/features/team-list'
import { QueryClient, QueryClientProvider } from 'react-query'
import { AuthContext } from '../../src/Auth/AuthContext'
import useTeams from '../../src/hooks/useTeams'

jest.mock('../../src/hooks/useTeams')

const mockNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}))

describe('TeamList component', () => {
  let queryClient: QueryClient

  const mockContext = {
    currentUser: {
      uid: '1234',
    },
  }

  beforeAll(() => {
    queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          retry: false,
        },
      },
    })
  })

  afterEach(() => {
    queryClient.clear()
  })

  it('renders loading message while teams are being fetched', async () => {
    useTeams.mockReturnValue({ isLoading: true })

    render(
      <QueryClientProvider client={queryClient}>
        <AuthContext.Provider value={mockContext}>
          <TeamList />
        </AuthContext.Provider>
      </QueryClientProvider>,
      { wrapper: BrowserRouter }
    )

    expect(await screen.findByText('Loading...')).toBeInTheDocument()
  })

  it('renderrs error message if teams fail to fetch', async () => {
    useTeams.mockReturnValue({ isError: true })

    render(
      <QueryClientProvider client={queryClient}>
        <AuthContext.Provider value={mockContext}>
          <TeamList />
        </AuthContext.Provider>
      </QueryClientProvider>,
      { wrapper: BrowserRouter }
    )

    expect(await screen.findByText('Error loading teams')).toBeInTheDocument()
  })

  it('renders team names after teams are fetched', async () => {
    const mockTeams = [
      { id: '1', teamName: 'Team A' },
      { id: '2', teamName: 'Team B' },
    ]
    useTeams.mockReturnValue({ data: mockTeams })

    render(
      <QueryClientProvider client={queryClient}>
        <AuthContext.Provider value={mockContext}>
          <TeamList />
        </AuthContext.Provider>
      </QueryClientProvider>,
      { wrapper: BrowserRouter }
    )

    expect(await screen.findByText('Team A')).toBeInTheDocument()
    expect(screen.getByText('Team B')).toBeInTheDocument()
  })

  it('renders nothing if teams is null', () => {
    const mockTeams = null
    useTeams.mockReturnValue({ data: mockTeams })
    render(
      <QueryClientProvider client={queryClient}>
        <AuthContext.Provider value={mockContext}>
          <TeamList />
        </AuthContext.Provider>
      </QueryClientProvider>,
      { wrapper: BrowserRouter }
    )
    const newTeamButton = screen.getByRole('button', { name: /new team/i })
    const divider = screen.getByRole('separator')
    expect(newTeamButton).toBeInTheDocument()
    expect(divider).toBeInTheDocument()
    expect(screen.queryByText(/error loading teams/i)).not.toBeInTheDocument()
    expect(screen.queryByText(/loading/i)).not.toBeInTheDocument()
  })

  it('navigates to add team page when "New Team" button is clicked', () => {
    render(
      <QueryClientProvider client={queryClient}>
        <AuthContext.Provider value={mockContext}>
          <TeamList />
        </AuthContext.Provider>
      </QueryClientProvider>,
      { wrapper: BrowserRouter }
    )
    const newTeamButton = screen.getByText('New Team')
    newTeamButton.click()
    expect(mockNavigate).toHaveBeenCalledWith('/add-team')
  })
})
