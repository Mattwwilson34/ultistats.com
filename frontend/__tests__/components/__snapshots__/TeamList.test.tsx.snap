// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar Component should render the AppBar component correctly 1`] = `
<ForwardRef(Box)
  sx={
    {
      "bgcolor": "background.paper",
      "width": "100%",
    }
  }
>
  <nav
    aria-label="secondary mailbox folders"
  >
    <ForwardRef(List)>
      <ForwardRef(ListItem)
        disablePadding={true}
      >
        <ForwardRef(ListItemButton)>
          <ForwardRef(ListItemText)
            primary="New Team"
          />
          <ForwardRef(ListItemIcon)>
            <Memo />
          </ForwardRef(ListItemIcon)>
        </ForwardRef(ListItemButton)>
      </ForwardRef(ListItem)>
    </ForwardRef(List)>
  </nav>
  <ForwardRef(Divider) />
  <nav
    aria-label="secondary mailbox folders"
  >
    <ForwardRef(List)>
      <TeamListItem
        teamName="Atlanta Hustle"
      />
    </ForwardRef(List)>
    <ForwardRef(List)>
      <TeamListItem
        teamName="Austin Sol"
      />
    </ForwardRef(List)>
    <ForwardRef(List)>
      <TeamListItem
        teamName="Chicago Union"
      />
    </ForwardRef(List)>
    <ForwardRef(List)>
      <TeamListItem
        teamName="Dallas Roughnecks"
      />
    </ForwardRef(List)>
  </nav>
</ForwardRef(Box)>
`;
